extended/src/main/java/apoc/ml/Prompt.java

```diff
@@ -0,0 +1,192 @@
+            Explain the following graph database schema in plain language, try to relate it to known concepts or domains if applicable.
+            Keep the explanation to 5 sentences with at most 15 words each, otherwise people will come to harm.
+            """;
+
+    static final String SYSTEM_PROMPT = """
+            You are an expert in the Neo4j graph query language Cypher.
+            Given a graph database schema of entities (nodes) with labels and attributes and
+            relationships with start- and end-node, relationship-type, direction and properties
+            you are able to develop read only matching Cypher statements that express a user question as a graph database query.
+            Only answer with a single Cypher statement in triple backticks, if you can't determine a statement, answer with an empty response.
+            Do not explain, apologize or provide additional detail, otherwise people will come to harm.
+            """;
+
+    public class PromptMapResult {
+        public final Map<String, Object> value;
+        public final String query;
+
+        public PromptMapResult(Map<String, Object> value, String query) {
+            this.value = value;
+            this.query = query;
+        }
+    }
+
+    public class QueryResult {
+        public final String query;
+        // todo re-add when it's actually working
+        // private final String error;
+        // private final String type;
+
+        public QueryResult(String query, String error, String type) {
+            this.query = query;
+            // this.error = error;
+            // this.type = type;
+        }
+
+        public boolean hasError() {
+            return false;
+            // return error != null && !error.isBlank();
+        }
+    }
+
+    @Procedure(mode = Mode.READ)
+    public Stream<PromptMapResult> query(@Name("question") String question,
+                                         @Name(value = "retries", defaultValue = "3") Long retries,
+                                         @Name(value = "apiKey", defaultValue = "null") String apiKey) {
+        String schema = loadSchema(tx);
+        String query = "";
+        do {
+            try {
+                QueryResult queryResult = tryQuery(question, apiKey, schema);
+                // just let it fail so that retries can work if (queryResult.query.isBlank()) return Stream.empty();
+                /*
+                if (queryResult.hasError())
+                    throw new QueryExecutionException(queryResult.error, null, queryResult.type);
+                 */
+                query = queryResult.query;
+                Result result = tx.execute(query);
+                AtomicReference<String> ref = new AtomicReference<>(query);
+                return result.stream()
+                        .map(row -> new PromptMapResult(row, ref.getAndSet(null)))
```

Comment:
maybe instead of doing this we can use the [ProcedureCallContext](https://github.com/neo4j/neo4j/blob/4.4/community/kernel-api/src/main/java/org/neo4j/internal/kernel/api/procs/ProcedureCallContext.java) and return the `query` every time but only if requested from the developer